rust2/core/src/options.rs

```diff
@@ -0,0 +1,492 @@
+        Self::Bytes(value.into())
+    }
+}
+
+/// ADBC revision versions.
+#[derive(Clone, Copy)]
+pub enum AdbcVersion {
+    /// Version 1.0.0.
+    V100,
+    /// Version 1.1.0.
+    V110,
+}
+
+impl From<AdbcVersion> for c_int {
+    fn from(value: AdbcVersion) -> Self {
+        match value {
+            AdbcVersion::V100 => constants::ADBC_VERSION_1_0_0,
+            AdbcVersion::V110 => constants::ADBC_VERSION_1_1_0,
+        }
+    }
+}
+
+/// Info codes for database/driver metadata.
+#[derive(Debug)]
+pub enum InfoCode {
+    /// The database vendor/product name (type: utf8).
+    VendorName,
+    /// The database vendor/product version (type: utf8).
+    VendorVersion,
+    /// The database vendor/product Arrow library version (type: utf8).
+    VendorArrowVersion,
+    /// The driver name (type: utf8).
+    DriverName,
+    /// The driver version (type: utf8).
+    DriverVersion,
+    /// The driver Arrow library version (type: utf8).
+    DriverArrowVersion,
+    /// The driver ADBC API version (type: int64).
+    ///
+    /// # Since
+    ///
+    /// ADBC API revision 1.1.0
+    DriverAdbcVersion,
+    // TODO(alexandreyc): add new codes (see https://github.com/apache/arrow-adbc/commit/aa04aadccd319e6fa3abb07154fa8d87b58d5c21)
+}
+
+impl From<&InfoCode> for u32 {
+    fn from(value: &InfoCode) -> Self {
+        match value {
+            InfoCode::VendorName => constants::ADBC_INFO_VENDOR_NAME,
+            InfoCode::VendorVersion => constants::ADBC_INFO_VENDOR_VERSION,
+            InfoCode::VendorArrowVersion => constants::ADBC_INFO_VENDOR_ARROW_VERSION,
+            InfoCode::DriverName => constants::ADBC_INFO_DRIVER_NAME,
+            InfoCode::DriverVersion => constants::ADBC_INFO_DRIVER_VERSION,
+            InfoCode::DriverArrowVersion => constants::ADBC_INFO_DRIVER_ARROW_VERSION,
+            InfoCode::DriverAdbcVersion => constants::ADBC_INFO_DRIVER_ADBC_VERSION,
+        }
+    }
+}
+
+impl TryFrom<u32> for InfoCode {
+    type Error = Error;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            constants::ADBC_INFO_VENDOR_NAME => Ok(InfoCode::VendorName),
+            constants::ADBC_INFO_VENDOR_VERSION => Ok(InfoCode::VendorVersion),
+            constants::ADBC_INFO_VENDOR_ARROW_VERSION => Ok(InfoCode::VendorArrowVersion),
+            constants::ADBC_INFO_DRIVER_NAME => Ok(InfoCode::DriverName),
+            constants::ADBC_INFO_DRIVER_VERSION => Ok(InfoCode::DriverVersion),
+            constants::ADBC_INFO_DRIVER_ARROW_VERSION => Ok(InfoCode::DriverArrowVersion),
+            constants::ADBC_INFO_DRIVER_ADBC_VERSION => Ok(InfoCode::DriverAdbcVersion),
+            v => Err(Error::with_message_and_status(
+                &format!("Unknown info code: {}", v),
+                Status::InvalidData,
+            )),
+        }
+    }
+}
+
+/// Depth parameter for [get_objects][crate::Connection::get_objects] method.
+#[derive(Debug)]
+pub enum ObjectDepth {
```

Comment:
I've chosen to not rely on integer representation of enums (and instead write explicit TryFrom/From) because of this:

```rust
use num_enum::{FromPrimitive, IntoPrimitive};

#[repr(u32)]
#[derive(FromPrimitive, IntoPrimitive)]
pub enum InfoCode {
    VendorName = 0,
    VendorVersion = 1,
    VendorArrowVersion = 2,
    DriverName = 100,
    DriverVersion = 101,
    DriverArrowVersion = 102,
    #[num_enum(catch_all)]
    Other(u32),
}

fn main() {
    println!("size = {}", std::mem::size_of::<InfoCode>());
    // size = 8
    // not u32!
}
```

This snippet of code is present is the current version (the one in `/rust`). When crossing FFI boundaries by passing for instance an `&[InfoCode]` it introduces a memory corruption...

So, I we stay away from `#[num_enum(catch_all)]` I'm OK to rely on `#[repr(integer)]` instead of defining constants and conversion traits.

Reply:
Well, after reconsideration here are my conclusions:

1. `Status` and `InfoCode` are represented as fixed-width integers in C (`u8` and `u32` respectively). In this case there should be no probleme using `#[repr(u8/u32)]`.

2. `ObjectDepth` and `AdbcVersion` are represented in C as `int` but we cannot `#[repr(c_int)]` so we must choose an `i*` but this is platform dependent... One solution would be to choose `i16` (because it's the minimum size imposed by the C standard for `int` and we will probably never have more than 2^16 variants) and widen values to `c_int` when required.

Personally I find this unecessary mental overhead and explicit conversion between "native" enums and integer is simpler (and it has the added benefit of handling invalid values as well).

Reply:
I would rather avoid making FFI so central, we should use native Rust conventions where possible. (Of course it is easier to structure things in a way that maps cleanly but we shouldn't force everything to be 1:1, it defeats the point.)

FWIW the Go definitions mostly do not make concessions like this to FFI considerations but FFI works fine.